### DISCUSSION TITLE
How to store and access JWT token in cookies

### DISCUSSION THREAD

**Discussion by Steveb599** [vercel/next.js#57596] [Help]
### Summary

I have a seperate backend which handles auth, which when I send a fetch request for log in or register, I receive a JWT token which I then need to send to the server on each request through authorisation. also I plan on having passportJS on the backend for social log in.
What is the best way to handle the jwt token, store them cookies and access them via client and server components? 
and how do I handle protected routes in this case?

**Comment by Kepron** [comment]
I too am trying to create an authentication system with a separate backend server. I've been researching this for a few days, but I haven't found a decent resource that addresses this situation. So I decided to do it myself, step by step. So far I have done this:

1. When creating a JWT token, I add the user id to the payload.
2. When creating the user session, I get the user id from the JWT payload and fetch the user data with this id from the database.

The code I used to do this on both the client-side and server-side is as follows;

Backend Login Api Function
```
async function authenticate(req, res) {
	...
	const { body } = req;
	try {
		...
		const user = {
			sub: userID,
		};
		const token = await new SignJWT(user)
			.setProtectedHeader({ alg: "HS256" })
			.setIssuedAt()
			.setExpirationTime("2h") // 1d
			.sign(getJwtSecretKey());
		...
		return res.status(200).json({ status: "ok", token });
	} catch (error) {
		...
	}
}
```

useSession.js (this for client-side)
```
import { getCookie } from "cookies-next";
import React from "react";
import { authAPI } from "src/_routes/auth.router";

export const useSession = () => {
	const [auth, setAuth] = React.useState(null);
	const getSessionData = async () => {
		const token = getCookie("token") ?? null;
		try {
			const result = await authAPI.getSession(token); // this contains axios code that calls the /api/session API.
			setAuth(result);
		} catch (error) {
			console.log(error);
		}
	};

	React.useEffect(() => {
		getSessionData();
	}, []);
	return auth;
};
```

useAuth.js (this for server-side)
```
"use server";
import { cookies } from "next/headers";
import { authAPI } from "src/_routes/auth.router";

export const useAuth = async () => {
	const cookieList = cookies();
	const { value: token } = cookieList.get("token") ?? { value: null };
	try {
		const result = await authAPI.getSession(token); // this contains axios code that calls the /api/session API.
		return result;
	} catch (error) {
		return null;
	}
};
```

But I have to warn you about something. In the  `catch` in the useAuth code, for some reason i dont know, I cannot delete the cookie if an error occurs. That's why I'm returning `null`.

/api/session has middleware on nodejs side.

```
app.use("/api/session", requireAuth, authRouter);
```

If any error is returned from /api/session, useAuth will throw null, but that's it, nothing more. To handle the error before it reaches useAuth, I check /api/session in middleware. So if an error is returned in /api/session, the first middleware will handle it. Since we can do things like delete cookies in middleware, it is easier to handle the situation.

```
import { NextResponse } from "next/server";
import { verifyJwtToken } from "src/lib/backend/verifyJwt";

export default async function middleware(request) {
	const { url, nextUrl, cookies } = request;
	const { value: token } = cookies.get("token") ?? { value: null };
	const hasVerifiedToken = token && (await verifyJwtToken(token));
	if (nextUrl.pathname.startsWith(`/login`)) {
		if (hasVerifiedToken) {
			return NextResponse.redirect(new URL("/dashboard", url));
		}
		return NextResponse.next();
	}
	if (!hasVerifiedToken && nextUrl.pathname.startsWith(`/dashboard`)) {
		const response = NextResponse.redirect(new URL(`/login?reason=expiredToken`, url));
		response.cookies.delete("token");
		return response;
	}
	if (hasVerifiedToken && nextUrl.pathname.startsWith(`/dashboard`)) {
		const fetchSession = await fetch(`${process.env.NEXT_PUBLIC_API_URL_V2}/v1/auth/session`, {
			method: "POST",
			body: JSON.stringify({ token }),
			headers: {
				"Content-Type": "application/json",
				Authorization: `Bearer ${token}`,
			},
		});
		const resultFetch = await fetchSession.json();

		if (resultFetch.status === "error") { // status is the custom value that I send from the nodejs server.
			const response = NextResponse.redirect(new URL(`/login?reason=${resultFetch.errorType}`, url)); // errorType is the custom value that I send from the nodejs server.
			response.cookies.delete("token");
			return response;
		}
	}
	return NextResponse.next();
}

export const config = {
	matcher: ["/((?!api|_next/static|_next/image|favicon.ico).*)"],
};
```

That's all I could think of so far. I hope this helps you. If anyone sees this post and has better ideas, please don't hesitate to let me know.
